internal/listtoken/list_token.go

```diff
@@ -0,0 +1,441 @@
+
+// NewPagination creates a new token with the pagination subtype.
+func NewPagination(
+	ctx context.Context,
+	createTime time.Time,
+	typ resource.Type,
+	grantsHash []byte,
+	lastItemId string,
+	lastItemCreateTime time.Time,
+) (*Token, error) {
+	const op = "listtoken.NewPagination"
+
+	if len(grantsHash) == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing grants hash")
+	}
+	if createTime.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "create time is in the future")
+	}
+	if createTime.Before(time.Now().AddDate(0, 0, -30)) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "create time is too old")
+	}
+	if lastItemId == "" {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing last item ID")
+	}
+	if lastItemCreateTime.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "last item create time is in the future")
+	}
+
+	return &Token{
+		CreateTime:   createTime,
+		ResourceType: typ,
+		GrantsHash:   grantsHash,
+		Subtype: &PaginationToken{
+			LastItemId:         lastItemId,
+			LastItemCreateTime: lastItemCreateTime,
+		},
+	}, nil
+}
+
+// NewStartRefresh creates a new token with a start-refresh subtype.
+func NewStartRefresh(
+	ctx context.Context,
+	createTime time.Time,
+	typ resource.Type,
+	grantsHash []byte,
+	previousDeletedIdsTime time.Time,
+	previousPhaseUpperBound time.Time,
+) (*Token, error) {
+	const op = "listtoken.NewStartRefresh"
+
+	if len(grantsHash) == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing grants hash")
+	}
+	if createTime.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "create time is in the future")
+	}
+	if createTime.Before(time.Now().AddDate(0, 0, -30)) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "create time is too old")
+	}
+	if previousDeletedIdsTime.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "previous deleted ids time is in the future")
+	}
+	if previousPhaseUpperBound.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "previous phase upper bound is in the future")
+	}
+
+	return &Token{
+		CreateTime:   createTime,
+		ResourceType: typ,
+		GrantsHash:   grantsHash,
+		Subtype: &StartRefreshToken{
+			PreviousPhaseUpperBound: previousPhaseUpperBound,
+			PreviousDeletedIdsTime:  previousDeletedIdsTime,
+		},
+	}, nil
+}
+
+// NewRefresh creates a new token with a refresh subtype.
+func NewRefresh(
+	ctx context.Context,
+	createTime time.Time,
+	typ resource.Type,
+	grantsHash []byte,
+	previousDeletedIdsTime time.Time,
+	phaseUpperBound time.Time,
+	phaseLowerBound time.Time,
+	lastItemId string,
+	lastItemUpdateTime time.Time,
+) (*Token, error) {
+	const op = "listtoken.NewRefresh"
+
+	if len(grantsHash) == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing grants hash")
+	}
+	if createTime.After(time.Now()) {
```

Comment:
do you want one value for now or several?

Reply:
See other answer